qiskit_experiments/data_processing/data_processor.py

```diff
@@ -0,0 +1,118 @@
+# that they have been altered from the originals.
+
+"""Actions done on the data to bring it in a usable form."""
+
+from typing import Any, Dict, List, Set, Tuple
+
+from qiskit_experiments.data_processing.data_action import DataAction
+from qiskit_experiments.data_processing.exceptions import DataProcessorError
+
+
+class DataProcessor:
+    """
+    A DataProcessor defines a sequence of operations to perform on experimental data.
+    Calling an instance of DataProcessor applies this sequence on the input argument.
+    A DataProcessor is created with a list of DataAction instances. Each DataAction
+    applies its _process method on the data and returns the processed data. The nodes
+    in the DataProcessor may also perform data validation and some minor formatting.
+    The output of one data action serves as input for the next data action.
+    DataProcessor.__call__(datum) usually takes in an entry from the data property of
+    an ExperimentData object (i.e. a dict containing metadata and memory keys and
+    possibly counts, like the Result.data property) and produces the formatted data.
+    DataProcessor.__call__(datum) extracts the data from the given datum under
+    DataProcessor._input_key (which is specified at initialization) of the given datum.
+    """
+
+    def __init__(self, input_key: str, data_actions: List[DataAction] = None):
+        """Create a chain of data processing actions.
+
+        Args:
+            input_key: The initial key in the datum Dict[str, Any] under which the data processor
+                will find the data to process.
+            data_actions: A list of data processing actions to construct this data processor with.
+                If None is given an empty DataProcessor will be created.
+        """
+        self._input_key = input_key
+        self._nodes = []
+
+        if data_actions:
+            for node in data_actions:
+                self.append(node)
+
+    def append(self, node: DataAction):
+        """
+        Append new data action node to this data processor.
+
+        Args:
+            node: A DataAction that will process the data.
+        """
+        self._nodes.append(node)
+
+    def __call__(self, datum: Dict[str, Any]) -> Tuple[Any, Any]:
+        """
+        Call self on the given datum. This method sequentially calls the stored data actions
+        on the datum.
+
+        Args:
+            datum: A single item of data, typically from an ExperimentData instance, that needs
+                to be processed. This dict also contains the metadata of each experiment.
+
+        Returns:
+            processed data: The data processed by the data processor.
+
+        Raises:
+            DataProcessorError: if no nodes are present.
+        """
+        if len(self._nodes) == 0:
+            raise DataProcessorError("Cannot call an empty data processor.")
+
+        datum_ = datum[self._input_key]
```

Comment:
Maybe catch the case that `datum` does not have a key named `self._input_key`, and produce a `ValueError` exception with a nice error message

Reply:
Good idea. Done here: d749d95